src/Test/Hspec/Yesod.hs

```diff
@@ -1573,26 +1561,23 @@ siteToYesodExampleData
     -> YesodExampleData site
 siteToYesodExampleData site =
     YesodExampleData
-        { yedApp = \_ -> toWaiAppPlain site
+        { yedCreateApplication = \site' middleware -> middleware <$> toWaiAppPlain site'
+        , yedMiddleware = id
         , yedSite = site
         , yedCookies = M.empty
         , yedResponse = Nothing
         , yedTestCleanup = pure ()
         }
 
-instance YesodDispatch site => Example (SIO (YesodExampleData site) a) where
-    type Arg (SIO (YesodExampleData site) a) = TestApp site
+instance Example (SIO (YesodExampleData site) a) where
+    type Arg (SIO (YesodExampleData site) a) = YesodExampleData site
```

Comment:
This is the main annoyance with ditching the `newtype`. Folks are *required* to migrate to `YesodExampleData site` instead of `TestApp site`. 

So, *with* the newtype, we have this:

```haskell
newtype YesodExample site a 

assertion :: (MonadState (YesodExampleData site) m, MonadIO m) => m ()

instance Example (SIO (YesodExampleData site) a)

instance Example (YesodExample site a)
```

And then you can work with `SIO` tests or `YesodExample` tests, and the `TestApp` vs `YesodExampleData` distinction is fine.

Without the newtype, then we have:

```diff
type YesodExample site a 

assertion 
-    :: (MonadState (YesodExampleData site) m, MonadIO m) 
-    => 
-    m ()
+    YesodExample site ()

 instance Example (SIO (YesodExampleData site) a where
    type Arg (SIO (YesodExampleData site) a) = 
-    TestApp site
+    YesodExampleData site

- instance Example (YesodExample site a)
```

This forces everyone to drop `TestApp` for `YesodExampleData` when using this, which makes `TestApp` entirely redundant.

Ultimately, I think this is fine. We can provide a compatibility shim for `(site, Middleware) -> YesodExampleData site` pretty easily, and an hspec hook for that.

The polymorphic assertions are an interesting case... I do wish `hspec`'s design didn't have the `Arg` type family causing the ambiguity issues, but it's not clear to me how else you'd do it.